nifi-nar-bundles/nifi-cdc/nifi-cdc-mssql-bundle/nifi-cdc-mssql-processors/src/main/java/org/apache/nifi/cdc/mssql/MSSQLCDCUtils.java

```diff
@@ -0,0 +1,286 @@
+            }
+        }
+
+        return cdcTables;
+    }
+
+    public List<ColumnDefinition> getCaptureColumns(Connection con, long objectId) throws SQLException, CDCException {
+        ArrayList<ColumnDefinition> tableColumns = new ArrayList<>();
+        try(final PreparedStatement st = con.prepareStatement(getLIST_TABLE_COLUMNS())){
+            st.setLong(1, objectId);
+
+            final ResultSet resultSet = st.executeQuery();
+            while (resultSet.next()) {
+                String columnName = resultSet.getString("column_name");
+                int columnId = resultSet.getInt("column_id");
+                String columnType = resultSet.getString("column_type");
+                int columnOrdinal = resultSet.getInt("column_ordinal");
+                int isColumnKey = resultSet.getInt("key");
+
+                int jdbcType = TranslateMSSQLTypeToJDBCTypes(columnType);
+
+                //get column list
+                MSSQLColumnDefinition col = new MSSQLColumnDefinition(jdbcType, columnName, columnOrdinal, isColumnKey==1);
+                tableColumns.add(col);
+            }
+        }
+
+        return tableColumns;
+    }
+
+    public String getSnapshotSelectStatement(MSSQLTableInfo tableInfo){
+        final StringBuilder sbQuery = new StringBuilder();
+
+        sbQuery.append("SELECT " + getCURRENT_TIMESTAMP() + " tran_begin_time");
+
+        sbQuery.append(_columnSplit);
+        sbQuery.append(getCURRENT_TIMESTAMP() + " \"tran_end_time\"");
+
+        sbQuery.append(_columnSplit);
+        sbQuery.append("0 trans_id");
+
+        sbQuery.append(_columnSplit);
+        sbQuery.append("0 start_lsn");
+
+        sbQuery.append(_columnSplit);
+        sbQuery.append("0 seqval");
+
+        sbQuery.append(_columnSplit);
+        sbQuery.append("2 operation");
+
+        sbQuery.append(_columnSplit);
+        sbQuery.append("0 update_mask");
+
+        for(ColumnDefinition col : tableInfo.getColumns()){
+            MSSQLColumnDefinition mssqlColumnDefinition = (MSSQLColumnDefinition)col;
+
+            sbQuery.append(_columnSplit);
+            sbQuery.append("\"" + mssqlColumnDefinition.getName() + "\"");
+        }
+
+        sbQuery.append(_columnSplit);
+        sbQuery.append(getCURRENT_TIMESTAMP() + " EXTRACT_TIME");
+        sbQuery.append("\n");
+        sbQuery.append("FROM " + tableInfo.getSourceSchemaName() + ".\""+ tableInfo.getSourceTableName() + "\"");
+
+        return sbQuery.toString();
+    }
+
+    public String getCDCSelectStatement(MSSQLTableInfo tableInfo, boolean includePreupdateValues, Timestamp maxTime){
+        final StringBuilder sbQuery = new StringBuilder();
+
+        sbQuery.append("SELECT t.tran_begin_time\n" +
+                ",t.tran_end_time \"tran_end_time\"\n" +
+                ",CAST(t.tran_id AS bigint) trans_id\n" +
+                ",CAST(\"o\".\"__$start_lsn\" AS bigint) start_lsn\n" +
+                ",CAST(\"o\".\"__$seqval\" AS bigint) seqval\n" +
```

Comment:
As per Microsoft, the value of seqval can be much larger than bigint. Can you please convert this to a higher value? The source data type is binary(10) which roughly translates to decimal(29,0).

Reply:
You are correct, that this is what the spec says. It's also the spec for start_lsn and end_lsn.
I'm not sure why I didn't go with Decimal originally, it's been to long since I wrote it :). But are you actually having an issue? BigInt has a max positive size of 9,223,372,036,854,775,807. That is a lot of transactions...

Reply:
Yeah, I do see issues and as the BigInt value crossed, I am getting negative values which is causing issues for syncing to our datalake

Reply:
Were these changes made? I am now facing this issue.

Reply:
Seems we can't convert directly from binary(10) to a decimal easily. Is there a java function that will make this easier?

Reply:
I was just researching this. BIGINT is actually bigger than decimal in MS SQL....?  It looks like BIGINT actually holds up to 2^63 (-1), where as DECIMAL(38) only holds up to 2^38 (-1).  So... unless I'm missing something I don't see a reason to move to DECIMAL.  In fact, based on research, I think BIGINT might be the best fit.

Reply:
I believe Decimal is 10^38 not 2^38